File: haskell/packages/numbers-utils/src/Numbers/Utils/Operations/FactorUtils.hs
module Numbers.Utils.Operations.FactorUtils
    ( factor
    , factorRandom
    , primeFactor
    , primeFactorRandom
    ) where

import Numeric.Natural (Natural)
import System.Random (randomRIO)

-- |
-- = @factor@
--
-- The function @factor@ is used to calculate the factors of a number.
--
-- __Arguments:__
--
-- * @num: Natural@ - The number to calculate the factors.
--
-- __Return:__
--
-- * @[Natural]@ - The factors of the number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FactorUtils (factor)
-- let factors = factor 25
-- putStrLn $ "Factor[25] = " ++ show factors
-- @
--
-- __Output:__
--
-- @
-- "Factor[25] = [1,5,25]"
-- @
factor :: Natural -> [Natural]
factor num = [x | x <- [1 .. num], num `mod` x == 0]

-- |
-- = @factorRandom@
--
-- The function @factorRandom@ is used to calculate the factors of a random
-- number.
--
-- __Return:__
--
-- * @String@ - The factors of the random number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FactorUtils (factorRandom)
-- factors <- factorRandom
-- putStrLn factors
-- @
--
-- __Output:__
--
-- @
-- "Factor[5] = [1,5]"
-- @
factorRandom :: IO String
factorRandom = do
    randomNum <- randomRIO (1, 500000 :: Integer)
    let factors = factor (fromIntegral randomNum)
    return $ "Factor[" ++ show randomNum ++ "] = " ++ show factors

-- |
-- = @primeFactor@
--
-- The function @primeFactor@ is used to calculate the prime factors of a
-- number.
--
-- __Arguments:__
--
-- * @num: Natural@ - The number to calculate the prime factors.
--
-- __Return:__
--
-- * @[Natural]@ - The prime factors of the number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FactorUtils (primeFactor)
-- let primeFactors = primeFactor 25
-- putStrLn $ "PFactor[25] = " ++ show primeFactors
-- @
--
-- __Output:__
--
-- @
-- "PFactor[25] = [5,5]"
-- @
primeFactor :: Natural -> [Natural]
primeFactor 0 = []
primeFactor 1 = []
primeFactor num = go num 2
    where
        go n divisor
            | divisor * divisor > n = [n]
            | n `mod` divisor == 0 = divisor : go (n `div` divisor) divisor
            | divisor == 2 = go n (divisor + 1)
            | otherwise = go n (divisor + 2)

-- |
-- = @primeFactorRandom@
--
-- The function @primeFactorRandom@ is used to calculate the prime factors of a
-- random number.
--
-- __Return:__
--
-- * @String@ - The prime factors of the random number.
--
-- __Example:__
--
-- @
-- import Numbers.Utils.Operations.FactorUtils (primeFactorRandom)
-- primeFactors <- primeFactorRandom
-- putStrLn primeFactors
-- @
--
-- __Output:__
--
-- @
-- "PFactor[20] = [2,2,5]"
-- @
primeFactorRandom :: IO String
primeFactorRandom = do
    randomNum <- randomRIO (1, 500000 :: Integer)
    let primeFactors = primeFactor (fromIntegral randomNum)
    return $ "PFactor[" ++ show randomNum ++ "] = " ++ show primeFactors

File: kotlin/subprojects/numbers-utils/bin/main/com/prog/numbers/utils/operations/FactorUtils.kt
package com.prog.numbers.utils.operations

import java.math.BigInteger
import kotlin.random.Random

object FactorUtils {
    /**
     * # `factor`
     *
     * This function calculates the factors of a given number.
     *
     * __Arguments:__
     *
     * * `num: BigInteger` - The number to calculate the factors
     *
     * __Return:__
     *
     * * `List<BigInteger>` - The factors of the given number
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.numbers.utils.operations.FactorUtils.factor
     * val factors = factor(BigInteger.valueOf(324))
     * println("$factors")
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "[1, 2, 3, 4, 6, 9, 12, 18, 27, 36, 54, 81, 108, 162, 324]"
     * ```
     */
    fun factor(num: BigInteger): List<BigInteger> {
        val factors = mutableListOf<BigInteger>()
        val max = num.toLong()

        for (i in 1..max) {
            val divisor = BigInteger.valueOf(i)
            if (num.mod(divisor) == BigInteger.ZERO) {
                factors.add(divisor)
            }
        }

        return factors
    }

    /**
     * # `factorRandom`
     *
     * This function generates a random number and calculates its factors.
     *
     * __Return:__
     *
     * * `String` - The factors of the random number
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.numbers.utils.operations.FactorUtils.factorRandom
     * val factors = factorRandom()
     * println(factors)
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "Factor[5] = [1, 5]"
     * ```
     */
    fun factorRandom(): String {
        val randomNum = Random.nextInt(1, 26)
        val factors = factor(BigInteger.valueOf(randomNum.toLong()))
        return "Factor[$randomNum] = $factors"
    }

    /**
     * # `primeFactor`
     *
     * This function calculates the prime factors of a given number.
     *
     * __Arguments:__
     *
     * * `num: BigInteger` - The number to calculate the prime factors
     *
     * __Return:__
     *
     * * `List<BigInteger>` - The prime factors of the given number
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.numbers.utils.operations.FactorUtils.primeFactor
     * val primeFactors = primeFactor(BigInteger.valueOf(324))
     * println("$primeFactors")
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "[2, 2, 3, 3, 3, 3]"
     * ```
     */
    fun primeFactor(num: BigInteger): List<BigInteger> {
        if (num == BigInteger.ZERO) {
            return emptyList()
        }

        var num = num
        val primeFactors = mutableListOf<BigInteger>()
        val two = BigInteger.valueOf(2)

        while (num.mod(two) == BigInteger.ZERO) {
            primeFactors.add(two)
            num = num.divide(two)
        }

        var i = 3L
        while (BigInteger.valueOf(i * i) <= num) {
            val divisor = BigInteger.valueOf(i)
            while (num.mod(divisor) == BigInteger.ZERO) {
                primeFactors.add(divisor)
                num = num.divide(divisor)
            }
            i += 2
        }

        if (num > BigInteger.ONE) {
            primeFactors.add(num)
        }

        return primeFactors
    }

    /**
     * # `primeFactorRandom`
     *
     * This function generates a random number and calculates its prime factors.
     *
     * __Return:__
     *
     * * `String` - The prime factors of the random number
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.numbers.utils.operations.FactorUtils.primeFactorRandom
     * val primeFactors = primeFactorRandom()
     * println(primeFactors)
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "PFactor[5] = [5]"
     * ```
     */
    fun primeFactorRandom(): String {
        val randomNum = Random.nextInt(1, 26)
        val primeFactors = primeFactor(BigInteger.valueOf(randomNum.toLong()))
        return "PFactor[$randomNum] = $primeFactors"
    }
}
